### PR TITLE
Update Falcon Role and enable for linux installs / uninstalls

### PR DESCRIPTION
**Pull request by redhatrises**
Get Falcon role working on RHEL and update the role to follow best practices for variables.

### CODE REVIEW

File: roles/falcon/defaults/main.yml [hunk 1]
```diff
@@ -1,83 +1,73 @@
...
-      #
-      # Which CrowdStrike API should we download from? Possible values:
-      #       US-GOV-1:   api.laggar.gcw.crowdstrike.com
-      #       EU-1:       api.eu-1.crowdstrike.com
-      #       US-2:       api.us-2.crowdstrike.com
-      #       Catch all:  api.crowdstrike.com
-      #
-      url: "api.crowdstrike.com"
+# If the installation method is 'url', provide the url for the sensor to
+# be downloaded.
```
Comment (shawndwells) [inline roles/falcon/defaults/main.yml:58]:
```suggestion
# be downloaded from.
```

File: roles/falcon/defaults/main.yml [hunk 2]
```diff
@@ -1,83 +1,73 @@
...
-    trace: ""
-    feature: ""
-    message_log: ""
-    billing: ""
-    assert: ""
-    memfail_grace_period: ""
-    memfail_every_n: ""
+# If using 'api' or 'url' methods, the time (in seconds) the installer
+# should wait between retries.
+falcon_delay: 3
```
Comment (shawndwells) [inline roles/falcon/defaults/main.yml:69]: FWIW these numbers were totally made up in the original code. Was more a reflection of impatience, not wanting to wait 10 seconds, before an error. If you think they should be changed then please do update these values! :)

File: roles/falcon/defaults/main.yml [hunk 3]
```diff
@@ -1,83 +1,73 @@
...
+
+# If true, uninstall a system
+# yes/no bool, default no
```
Comment (shawndwells) [inline roles/falcon/defaults/main.yml:72]:
```suggestion
# yes/no bool, default false
```

File: roles/falcon/tasks/api.yml [hunk 4]
```diff
@@ -1,82 +1,57 @@
...
-#--silent \
-#--url "https://$target_cloud_api/oauth2/token" \
-#--header "content-type: application/x-www-form-urlencoded" \
-#--data client_id=$CS_API_CLIENT_ID \
-#--data client_secret=$CS_API_CLIENT_SECRET | \
-#  python3 -c "import sys, json; print(json.load(sys.stdin)['access_token'])")
-
-- name: CrowdStrike Falcon Installer | API Authorization Output
-  debug:
-    msg: "{{ crowdstrike.oauth2_token.json }}"
```
Comment (shawndwells) [inline roles/falcon/tasks/api.yml:22]: Should we keep the debug?

File: roles/falcon/tasks/api.yml [hunk 5]
```diff
@@ -1,82 +1,57 @@
...
-#echo "Target CID: $target_cid"
+      authorization: "Bearer {{ falcon_api_oauth2_token.json.access_token }}"
+      Content-Type: application/json
+  register: falcon_api_target_cid
 
-- name: CrowdStrike Falcon Installer | Detect SHA256 of Latest Linux Sensor
+- name: CrowdStrike Falcon | Detect SHA256 of Latest Linux Sensor
   uri:
-    url: https://{{ crowdstrike.api.url }}/sensors/combined/installers/v1?filter=platform%3A%22linux*%22%2Bos%3A%22$target_os%22&=&sort=version.desc&limit=1
+    url: "https://{{ falcon_api_url }}/sensors/combined/installers/v1?filter=platform%3A%22{{ falcon_os_family }}*%22%2Bos%3A%22{{ falcon_target_os }}%22&=&sort=version.desc&limit=1"
```
Comment (shawndwells) [inline roles/falcon/tasks/api.yml:27]: Clever, and good call!

**Review by shawndwells — CHANGES_REQUESTED** [review]

File: roles/falcon/defaults/main.yml (hunk 2 above)
Comment (redhatrises) [reply roles/falcon/defaults/main.yml:69]: @shawndwells they actually seem reasonable to me.

File: roles/falcon/tasks/api.yml (hunk 4 above)
Comment (redhatrises) [reply roles/falcon/tasks/api.yml:22]: @shawndwells why? No need to print out the OAuth2 Token to output.

File: roles/falcon/tasks/main.yml [hunk 6]
```diff
@@ -1,31 +1,31 @@
 ---
 # tasks file for crowdstrike-falcon-role
+- block:
+  - include_tasks: uninstall.yml
+  when: falcon_uninstall
+  become: true
+  become_user: root
```
Comment (shawndwells) [inline roles/falcon/tasks/main.yml:7]: Wayyy cleaner approach. Nice!

File: roles/falcon/tasks/preinstall.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    when: not falcon_api_client_secret
+
+  when: falcon_install_method == "api"
+
+- name: CrowdStrike Falcon | Configure Python Interpreter for older Linux OSes
+  set_fact:
+    ansible_python_interpreter: /usr/bin/python
+  when: ansible_distribution == "Amazon"
+
+- name: CrowdStrike Falcon | Determine if Target OS Is Enterprise Linux
```
Comment (shawndwells) [inline roles/falcon/tasks/preinstall.yml:25]:
```suggestion
- name: CrowdStrike Falcon | Determine if Endpoint Operating System Is RHEL, CentOS, or Oracle Linux
```

File: roles/falcon/tasks/preinstall.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  set_fact:
+     falcon_target_os: RHEL/CentOS/Oracle
+     falcon_os_family: "{{ ansible_system | lower }}"
+  when: ansible_os_family == "RedHat"
+
+- name: CrowdStrike Falcon | Determine if Target OS Is Amazon Linux
+  set_fact:
+    falcon_target_os: "Amazon%20Linux"
```
Comment (shawndwells) [inline roles/falcon/tasks/preinstall.yml:33]: Is ``%20`` needed for strings?

File: roles/falcon/tasks/preinstall.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    falcon_os_family: "{{ ansible_system | lower }}"
+  when: ansible_distribution == "Amazon"
+
+- name: "CrowdStrike Falcon | Determine if Target OS Is {{ ansible_distribution }}"
```
Comment (shawndwells) [inline roles/falcon/tasks/preinstall.yml:37]:
```suggestion
- name: "CrowdStrike Falcon | Endpoint Operating System Detected Is {{ ansible_distribution }}"
```

**Pull request merged**